lib/config/config.js

```diff
@@ -0,0 +1,285 @@
+                throw new TypeError(`Cannot serialize key "${key}" in ${objectKey}: Function values are not supported.`);
+            }
+
+        }
+
+        result[key] = value;
+    }
+
+    return result;
+}
+
+/**
+ * Normalizes the rules configuration. Ensure that each rule config is
+ * an array and that the severity is a number. This function modifies the
+ * rulesConfig.
+ * @param {Record<string, any>} rulesConfig The rules configuration to normalize.
+ * @returns {void}
+ */
+function normalizeRulesConfig(rulesConfig) {
+
+    for (const [ruleId, ruleConfig] of Object.entries(rulesConfig)) {
+
+        // ensure rule config is an array
+        if (!Array.isArray(ruleConfig)) {
+            rulesConfig[ruleId] = [ruleConfig];
+        }
+
+        // normalize severity
+        rulesConfig[ruleId][0] = severities.get(rulesConfig[ruleId][0]);
+    }
+
+}
+
+
+//-----------------------------------------------------------------------------
+// Exports
+//-----------------------------------------------------------------------------
+
+/**
+ * Represents a normalized configuration object.
+ */
+class Config {
+
+    /**
+     * The name to use for the language when serializing to JSON.
+     * @type {string|undefined}
+     */
+    #languageName;
+
+    /**
+     * The name to use for the processor when serializing to JSON.
+     * @type {string|undefined}
+     */
+    #processorName;
+
+    /**
+     * Creates a new instance.
+     * @param {Object} config The configuration object.
+     */
+    constructor(config) {
+
+        const { plugins, language, languageOptions, processor, ...otherKeys } = config;
+
+        // Validate config object
+        const schema = new ObjectSchema(flatConfigSchema);
+
+        schema.validate(config);
+
+        // first, copy all the other keys over
+        Object.assign(this, otherKeys);
+
+        // ensure that a language is specified
+        if (!language) {
+            throw new TypeError("Key 'language' is required.");
+        }
+
+        // copy the rest over
+        this.plugins = plugins;
+        this.language = language;
+
+        if (languageOptions) {
+            this.languageOptions = languageOptions;
+        }
+
+        // Check language value
+        if (typeof language === "string") {
+            const { pluginName, objectName: localLanguageName } = splitPluginIdentifier(language);
+
+            this.#languageName = language;
+
+            if (!plugins || !plugins[pluginName] || !plugins[pluginName].languages || !plugins[pluginName].languages[localLanguageName]) {
+                throw new TypeError(`Key "language": Could not find "${localLanguageName}" in plugin "${pluginName}".`);
+            }
+
+            this.language = plugins[pluginName].languages[localLanguageName];
+        } else if (typeof language === "object") {
+            this.#languageName = getObjectId(language);
+            this.language = language;
```

Comment:
Schema doesn't allow objects here, just `"pluginName/objectName"` strings.

https://github.com/eslint/eslint/blob/4840930b9d8b6aa3578fe234180425e9060ceeca/lib/config/flat-config-schema.js#L354-L357

Reply:
I know, but this isn't a raw config, it's a normalized config. We are already doing this normalization inside of [`FlatConfigArray#finalizeConfig`](https://github.com/eslint/eslint/pull/18763/files#diff-891c5261db14a885f6a80ecc03f7b50f01939c66a798c14b1292621def227e82L342), I'm just moving it into this class.

Reply:
I meant, at this point it can't happen that `language` variable is an object. `schema.validate(config);` would have already thrown an error on line 177.

Reply:
Ah gotcha. What you're saying is we don't need to check if language is an object at this point.

Reply:
I just copied this directly from `FlatConfigArray`, which was doing the same unnecessary check. So this will clean up and existing issue.